tls/s2n_post_handshake.c

```diff
@@ -13,66 +13,172 @@
-            default:
-                /* Ignore all other messages */
-                break;
+    switch (message_type) {
+        case TLS_KEY_UPDATE:
+            RESULT_GUARD_POSIX(s2n_key_update_recv(conn, in));
+            break;
+        case TLS_SERVER_NEW_SESSION_TICKET:
+            RESULT_GUARD(s2n_tls13_server_nst_recv(conn, in));
+            break;
+        case TLS_HELLO_REQUEST:
+            RESULT_GUARD(s2n_client_hello_request_recv(conn));
+            break;
+        default:
+            /* Ignore unknown messages */
+            RESULT_ENSURE(!s2n_post_handshake_is_known(message_type), S2N_ERR_BAD_MESSAGE);
+            break;
+    }
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_post_handshake_message_recv(struct s2n_connection *conn)
+{
+    RESULT_ENSURE_REF(conn);
+
+    struct s2n_stuffer *in = &conn->in;
+    struct s2n_stuffer *message = &conn->post_handshake.in;
+    uint8_t message_type = 0;
+    uint32_t message_len = 0;
+
+    /* We always start reading from the beginning of the message */
+    RESULT_GUARD_POSIX(s2n_stuffer_reread(message));
+
+    /* If no space for the fragment exists, start with the minimum, static space */
+    if (s2n_stuffer_is_freed(message)) {
+        struct s2n_blob b = { 0 };
+        RESULT_GUARD_POSIX(s2n_blob_init(&b, conn->post_handshake.in_bytes,
+                sizeof(conn->post_handshake.in_bytes)));
+        RESULT_GUARD_POSIX(s2n_stuffer_init(message, &b));
+    }
+
+    /* Try to read the header */
+    if (s2n_stuffer_data_available(message) < TLS_HANDSHAKE_HEADER_LENGTH) {
+        uint32_t remaining = TLS_HANDSHAKE_HEADER_LENGTH - s2n_stuffer_data_available(message);
+        uint32_t to_read = MIN(remaining, s2n_stuffer_data_available(in));
+        RESULT_GUARD_POSIX(s2n_stuffer_copy(in, message, to_read));
+    }
+    RESULT_ENSURE(s2n_stuffer_data_available(message) >= TLS_HANDSHAKE_HEADER_LENGTH, S2N_ERR_IO_BLOCKED);
+
+    /* Parse the header */
+    RESULT_GUARD(s2n_handshake_parse_header(message, &message_type, &message_len));
+    RESULT_ENSURE(message_len == 0 || s2n_stuffer_data_available(in), S2N_ERR_IO_BLOCKED);
+
+    /* If the message is not fragmented, just process it directly from conn->in.
+     * This will be the most common case, and does not require us to allocate
+     * any new memory.
+     */
+    if (s2n_stuffer_data_available(message) == 0 && s2n_stuffer_data_available(in) >= message_len) {
+        struct s2n_stuffer full_message = { 0 };
+        struct s2n_blob full_message_blob = { 0 };
+        RESULT_GUARD_POSIX(s2n_blob_init(&full_message_blob, s2n_stuffer_raw_read(in, message_len), message_len));
+        RESULT_GUARD_POSIX(s2n_stuffer_init(&full_message, &full_message_blob));
+        RESULT_GUARD_POSIX(s2n_stuffer_skip_write(&full_message, message_len));
+        RESULT_GUARD(s2n_post_handshake_process(conn, &full_message, message_type));
+        return S2N_RESULT_OK;
+    }
+
+    /* Skip unknown message types.
```

Comment:
/ ** Check that the message type is an acceptable post-handshake message. We don't want to allocate memory for the remaining fragment if we won't actually parse the message. **/